### PR TITLE
Extended ds download command with specific data structure and version

### PR DESCRIPTION
**Pull request by ivan-afanasiev**
# 🚀 Add Version-Specific Data Structure Download Functionality

## 📋 Summary

This PR extends the `snowplow-cli ds download` command with comprehensive version-specific download capabilities, allowing users to download specific versions, all versions, or filter by environment. The implementation includes proper file naming conventions, robust error handling, and maintains full backward compatibility.

### 🎯 Core Functionality
- **Specific Version Download**: Download a particular version of a data structure
- **All Versions Download**: Download all available versions of a data structure
- **Environment Filtering**: Filter deployments by environment (DEV/PROD)
- **Smart File Naming**: Automatic version suffix inclusion in filenames
- **Backward Compatibility**: Existing functionality remains unchanged

### 🔧 New Command-Line Flags
```bash
--vendor string     # Vendor of the specific data structure
--name string       # Name of the specific data structure  
--format string     # Format of the data structure (default: jsonschema)
--version string    # Specific version to download
--all-versions      # Download all versions (mutually exclusive with --version)
--env string        # Filter by environment (DEV, PROD)
```

### 1. Download Specific Version
```bash
# Download shop entity version 2-0-0
snowplow-cli ds download --vendor com.hellofresh --name shop --format jsonschema --version 2-0-0
# Creates: com.hellofresh/shop_2-0-0.yaml
```

### 2. Download All Versions
```bash
# Download all versions of shop entity
snowplow-cli ds download --vendor com.hellofresh --name shop --format jsonschema --all-versions
# Creates: com.hellofresh/shop_1-0-0.yaml, com.hellofresh/shop_2-0-0.yaml
```

### 3. Environment-Specific Downloads
```bash
# Download only production versions
snowplow-cli ds download --vendor com.hellofresh --name shop --format jsonschema --all-versions --env PROD
```

### 4. Latest Version (Existing Behavior)
```bash
# Download latest version (no version suffix)
snowplow-cli ds download --vendor com.hellofresh --name shop --format jsonschema
# Creates: com.hellofresh/shop.yaml
```

#### Core Logic
- **`cmd/ds/download.go`**: Extended command with new flags and download logic
- **`internal/console/requests_ds.go`**: Added version-specific API functions
- **`internal/util/files.go`**: Enhanced file creation with version naming

#### New Functions Added
```go
// Hash generation for data structure identification
GenerateDataStructureHash(orgId, vendor, name, format string) string

// Version-specific downloads
GetSpecificDataStructureVersion(ctx, client, dsHash, version string) (*DataStructure, error)
GetAllDataStructureVersions(ctx, client, dsHash, envFilter string) ([]DataStructure, error)

// Enhanced file creation with version support
CreateDataStructuresWithVersions(dss []DataStructure, isPlain bool, includeVersions bool) error
```

### 🔄 API Integration

The implementation uses a **two-step API approach** for reliable version retrieval:

1. **Listing API**: `/data-structures/v1/{hash}` - Get metadata and deployments
2. **Versions API**: `/data-structures/v1/schemas/versions` - Get actual schema content

This approach ensures we can retrieve any version, not just the latest, by matching the requested version from the comprehensive versions list.

### 🎨 Smart File Naming

The system automatically determines when to include version suffixes:

| Download Type | Filename Pattern | Example |
|---------------|------------------|---------|
| Specific Version | `{name}_{version}.{ext}` | `shop_2-0-0.yaml` |
| All Versions | `{name}_{version}.{ext}` | `shop_1-0-0.yaml`, `shop_2-0-0.yaml` |
| Latest Version | `{name}.{ext}` | `shop.yaml` |
| Bulk Download | `{name}.{ext}` | `shop.yaml` |

### 📊 Test Coverage
- **94 total tests** with **100% pass rate**
- **15 new test cases** specifically for version functionality
- **3 new test files** created for comprehensive coverage

#### 1. File Operations (`internal/util`)
```bash
✅ TestCreateDataStructuresWithVersions_IncludeVersionsTrue
✅ TestCreateDataStructuresWithVersions_IncludeVersionsFalse
✅ TestCreateDataStructuresWithVersions_MultipleVersionsSameName
✅ TestCreateDataStructuresWithVersions_JsonFormat
✅ TestCreateDataStructuresWithVersions_BackwardCompatibility
✅ TestCreateDataStructuresWithVersions_ComplexVersionNames
```

#### 2. API Functions (`internal/console`)
```bash
✅ TestGetSpecificDataStructureVersion_Success
✅ TestGetSpecificDataStructureVersion_VersionNotFound
✅ TestGetAllDataStructureVersions_Success
✅ TestGetAllDataStructureVersions_WithEnvironmentFilter
```

#### 3. Command Logic (`cmd/ds`)
```bash
✅ TestDownloadCommand_FlagValidation (5 sub-tests)
✅ TestDownloadCommand_IncludeVersionsLogic (4 sub-tests)
```

### 🔧 Test Features
- **Mock HTTP servers** with realistic API responses
- **Temporary file systems** for isolated testing
- **Error scenario coverage** including network failures
- **Edge case handling** for complex version numbers
- **Backward compatibility verification**

### ✅ Input Validation
- **Mutual exclusivity**: `--version` and `--all-versions` cannot be used together
- **Required parameters**: `--vendor`, `--name`, and `--format` must be provided together
- **Environment validation**: `--env` accepts only DEV/PROD values

### 🚨 Error Scenarios Handled
- **Non-existent versions**: Clear error messages with version information
- **API failures**: Graceful degradation with informative logging
- **Network issues**: Proper timeout and retry handling
- **Invalid responses**: Robust JSON parsing with fallback behavior

### 📝 User-Friendly Messages
```bash
# Clear error for missing version
❌ Schema data not found for version 3-0-0

# Informative success messages
✅ Downloaded specific version: vendor=com.hellofresh name=shop version=2-0-0
✅ Downloaded all versions: vendor=com.hellofresh name=shop count=3 env=PROD
```

### ✅ Existing Functionality Preserved
- **Bulk downloads** continue to work exactly as before
- **Latest version downloads** maintain original behavior
- **All existing flags** remain functional
- **File naming** unchanged for existing use cases

### 📖 Help Text Enhanced
```bash
$ snowplow-cli ds download --help

Download data structures from BDP Console.

By default, downloads the latest versions of all data structures from your development environment.

You can download specific data structures using --vendor, --name, and --format flags.
You can also download a specific version using --version flag, or all versions using --all-versions flag.
Use --env flag to filter deployments by environment (DEV, PROD).

Examples:
  # Download a specific data structure
  $ snowplow-cli ds download --vendor com.example --name login_click --format jsonschema

  # Download a specific version
  $ snowplow-cli ds download --vendor com.example --name login_click --format jsonschema --version 1-0-0

  # Download all versions
  $ snowplow-cli ds download --vendor com.example --name login_click --format jsonschema --all-versions

  # Download only production deployments
  $ snowplow-cli ds download --vendor com.example --name login_click --format jsonschema --all-versions --env PROD
```

### CODE REVIEW

File: cmd/ds/download.go [hunk 1]
```diff
@@ -64,12 +91,56 @@ Use --include-legacy to include them (they will be set to 'entity' schemaType).`
...
+		var dss []model.DataStructure
+
+		// Check if we're downloading a specific data structure
+		var includeVersions bool
+		if vendor != "" && name != "" && formatFlag != "" {
+			// Validate mutually exclusive flags
+			if version != "" && allVersions {
+				snplog.LogFatalMsg("validation error", fmt.Errorf("--version and --all-versions are mutually exclusive"))
+			}
+
```
Comment (gleb-lobov) [inline cmd/ds/download.go:103]: This can be expressed with cobra's [MarkFlagsMutuallyExclusive](https://pkg.go.dev/github.com/spf13/cobra#Command.MarkFlagsMutuallyExclusive)

File: cmd/ds/download.go [hunk 2]
```diff
@@ -85,4 +156,12 @@ func init() {
 	downloadCmd.PersistentFlags().StringArrayP("match", "", []string{}, "Match for specific data structure to download (eg. --match com.example/event_name or --match com.example)")
 	downloadCmd.PersistentFlags().Bool("include-legacy", false, "Include legacy data structures with empty schemaType (will be set to 'entity')")
 	downloadCmd.PersistentFlags().Bool("plain", false, "Don't include any comments in yaml files")
+
+	// New flags for specific data structure download
+	downloadCmd.PersistentFlags().String("vendor", "", "Vendor of the specific data structure to download (requires --name and --format)")
```
Comment (gleb-lobov) [inline cmd/ds/download.go:161]: >requires --name and --format

This should be expressed as [MarkFlagsRequiredTogether](https://pkg.go.dev/github.com/spf13/cobra#Command.MarkFlagsRequiredTogether)

File: internal/console/requests_ds.go [hunk 3]
```diff
@@ -466,3 +462,196 @@ func patchMeta(cnx context.Context, client *ApiClient, ds *model.DataStructureSe
 
 	return nil
 }
+
+// GenerateDataStructureHash generates a SHA-256 hash for a data structure
+// based on organization ID, vendor, name, and format as per Snowplow API documentation
+func GenerateDataStructureHash(orgId, vendor, name, format string) string {
```
Comment (gleb-lobov) [inline internal/console/requests_ds.go:468]: The logic is correct, but it's done differently on line 426. Either both should use this function, or none

File: cmd/ds/download_version_test.go [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+			tempDir := t.TempDir()
+			args := append(tt.args, tempDir)
+
+			// Create a new command for each test to avoid state pollution
+			cmd := downloadCmd
+
+			// Set up the command
+			cmd.SetArgs(args)
+
+			// Execute the command
```
Comment (gibbok-snowplow) [inline cmd/ds/download_version_test.go:76]: nit: same for these comments. thanks!

**Comment by ivan-afanasiev** [comment]
> Hi @ivan-afanasiev Thank for your contribution
> 
> Could you please describe the use-case that you have for the single version download? Im not against implementing it, but so far this tool was build to assist with git-ops, where your versions would be stored in the git history. I understand that there is a case when you first switch to git and don't have the history, but I don't see how the single version download helps there

Hi @gleb-lobov — thanks for the quick review!

My use case is a bit different: I'm building a code-generation tool for Snowplow event specifications because [Snowtype](https://docs.snowplow.io/docs/data-product-studio/snowtype/) doesn't match our workflow. Specifications can reference specific versions of events and entities, so I need a way to fetch a single version of a particular data structure so the generator can produce code for that exact schema. While calling [the Console API](https://console.snowplowanalytics.com/api/msc/v1/docs/) directly is possible, I thought it makes sense to add such kind of functionality into the CLI to simplify our code generator logic. 

I'm not a Go expert, so please feel free to challenge the implementation and suggest improvements — happy to iterate.

Thanks again!

**Comment by gibbok-snowplow** [comment]
@ivan-afanasiev Thanks for your message.

Regarding this point:

> I need a way to fetch a single version of a particular data structure so the generator can produce code for that exact schema.

In Snowtype, you can target specific data structures (including their exact versions) by defining them in your snowtype.config.json file. For example:

```json
{
  "dataStructures": [
    "xxx/aaa/jsonschema/3-0-1",
    "xxx/aaa/jsonschema/3-0-2"
  ]
}
```

I hope it helps.
